### ISSUE TITLE
Error when using --default-remote-provider and arrays of files as rule input

### ISSUE THREAD

**Issue by brand-fabian** [snakemake/snakemake#489]
Hello,

I encountered a bug when running snakemake with the default-remote-{provider,prefix} options and no-shared-fs. Please consider the information below.

**Snakemake version**
5.20.1

**Describe the bug**

When executing a step in the workflow which gets its input as a `dict(str -> list)`, the default-remote-{provider,prefix} options get applied to the list object not to the object in the list.

Assume this rule:

    rule name:
        input: unpack(my_func)
        output: "out.txt"
        ...
        wrapper: "file:wrapper/call_variants" # just an example

With the following input function:

    def my_func(wildcards):
        # Find sample information and retrieve files
        return {
             'variants': [ 'file1', 'file2', ... ],
             'tbi': [ 'tbi1', 'tbi2', ... ],
             'pedigree': 'file.ped'
        }

When calling the workflow with the options `--default-remote-provider S3 --default-remote-prefix test`, snakemake resolves the filenames to:

    `s3://test/[ 'file1', 'file2', ...]`

And throws an error that those files could not be found.

**Additional context**

This error is most likely generated in [rules.py::Rules::apply_default_remote](https://github.com/snakemake/snakemake/blob/7dd7d8b1cad48f9369deb42b54be3d52fcd65dd9/snakemake/rules.py#L434). When calling this function with the aforementioned `dict(str -> list)` instance it only iterates over the first level, calling the function `rules.py::Rules::apply_default_remote::apply` with the `list` as argument.

A possible fix is to change to `apply` function to:

            if (
                not is_flagged(value, "remote_object")
                and not is_flagged(value, "local")
                and not is_annotated_callable(value)
                and self.workflow.default_remote_provider is not None
            ):
                if isinstance(value, collections.abc.Iterable) and not isinstance(value, str):
                    return [ self.workflow.apply_default_remote(v) for v in value ]
                return self.workflow.apply_default_remote(value)
            else:
                return self.workflow.apply_default_remote(value)

If desired, I can submit a PR for this change.

**Comment by cademirch** [comment]
@brand-fabian Did this fix work for you? I am encountering an identical problem.

**Comment by brand-fabian** [comment]
Yes, changing the specific Lines worked for me. I have been using this changed snakemake build all the time. I dont know however, if these lines are still present in that form in current versions of the library.
